com.unity.multiplayer.mlapi/Runtime/Core/NetworkUpdateLoop.cs

```diff
@@ -0,0 +1,525 @@
+    /// Defines the required interface of a network update system being executed by the network update loop.
+    /// </summary>
+    public interface INetworkUpdateSystem
+    {
+        void NetworkUpdate(NetworkUpdateStage updateStage);
+    }
+
+    /// <summary>
+    /// Defines network update stages being executed by the network update loop.
+    /// </summary>
+    public enum NetworkUpdateStage : byte
+    {
+        Initialization = 1,
+        EarlyUpdate = 2,
+        FixedUpdate = 3,
+        PreUpdate = 4,
+        Update = 0,
+        PreLateUpdate = 5,
+        PostLateUpdate = 6
+    }
+
+    /// <summary>
+    /// Represents the network update loop injected into low-level player loop in Unity.
+    /// </summary>
+    public static class NetworkUpdateLoop
+    {
+        [RuntimeInitializeOnLoadMethod]
+        private static void Initialize()
+        {
+            var customPlayerLoop = PlayerLoop.GetCurrentPlayerLoop();
+
+            for (int i = 0; i < customPlayerLoop.subSystemList.Length; i++)
+            {
+                var playerLoopSystem = customPlayerLoop.subSystemList[i];
+
+                if (playerLoopSystem.type == typeof(Initialization))
+                {
+                    var subsystems = playerLoopSystem.subSystemList.ToList();
+                    {
+                        // insert at the bottom of `Initialization`
+                        subsystems.Add(NetworkInitialization.CreateLoopSystem());
+                    }
+                    playerLoopSystem.subSystemList = subsystems.ToArray();
+                }
+                else if (playerLoopSystem.type == typeof(EarlyUpdate))
+                {
+                    var subsystems = playerLoopSystem.subSystemList.ToList();
+                    {
+                        int indexOfScriptRunDelayedStartupFrame = subsystems.FindIndex(s => s.type == typeof(EarlyUpdate.ScriptRunDelayedStartupFrame));
+                        if (indexOfScriptRunDelayedStartupFrame < 0)
+                        {
+                            Debug.LogError($"{nameof(NetworkUpdateLoop)}.{nameof(Initialize)}: Cannot find index of `{nameof(EarlyUpdate.ScriptRunDelayedStartupFrame)}` loop system in `{nameof(EarlyUpdate)}`'s subsystem list!");
+                            return;
+                        }
+
+                        // insert before `EarlyUpdate.ScriptRunDelayedStartupFrame`
+                        subsystems.Insert(indexOfScriptRunDelayedStartupFrame, NetworkEarlyUpdate.CreateLoopSystem());
+                    }
+                    playerLoopSystem.subSystemList = subsystems.ToArray();
+                }
+                else if (playerLoopSystem.type == typeof(FixedUpdate))
+                {
+                    var subsystems = playerLoopSystem.subSystemList.ToList();
+                    {
+                        int indexOfScriptRunBehaviourFixedUpdate = subsystems.FindIndex(s => s.type == typeof(FixedUpdate.ScriptRunBehaviourFixedUpdate));
```

Comment:
Lines 72 to 84 mostly duplicates lines56 to 68. This is an anti-pattern. Could we have a single block that uses variables/parameters?
Same for lines 88 to 100, 104 to 116, 120 to 132, and 136 to 148.

Reply:
this pattern is necessary:

```
var subsystems = playerLoopSystem.subSystemList.ToList();
...
playerLoopSystem.subSystemList = subsystems.ToArray();
```

we can't reuse existing `List` instance and/or set it to subSystemsList as a static array.